Develop

AlexMAK1 opened the pull request:
begining

AlexMAK1 committed:
beginning

AlexMAK1 committed:
beginning

AlexMAK1 committed:
fixed bugs after checkstyle checking

AlexMAK1 committed:
fixed bugs after Github checking

AlexMAK1 committed:
Update README.md

add links to pull pequest

krylov-sergey-sergeevich commented on main_service/Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM amazoncorretto:11-alpine-jdk
+COPY target/*.jar app.jar
```
Можно еще вот так:
```
ARG JAR_FILE=target/*.jar
COPY ${JAR_FILE} app.jar
```

AlexMAK1 replied:
ок. в чем преимущество такого подхода?

krylov-sergey-sergeevich replied:
По сути прямых плюсов нет, скорее тут говорю о том что такая возможность есть =)

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/categories/CategoryController.java:
```diff
@@ -0,0 +1,53 @@
+
+@Slf4j
+@RestController
+public class CategoryController {
```
Можно накинуть аннотацию ломбока и избавиться от конструктора

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/categories/CategoryController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @GetMapping("/categories")
+    public List<CategoryDto> findAll(@RequestParam(name = "from", defaultValue = "0")
```
как то checkstyle поехал, переменные съехали по строкам

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/categories/CategoryConverter.java:
```diff
@@ -0,0 +1,28 @@
+import ru.practicum.main_service.categories.model.Category;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class CategoryConverter {
```
Отлично, можете ещё посмотреть в сторону библиотеки MapStruct, она сама мапит поля с одинаковым типом и именованием. https://www.baeldung.com/mapstruct

Применить только у вас не получится. Т.к. тесты не пройдут.
Если будут вопросы как использовать данную библиотеку - можете посмотреть это видео
https://youtu.be/Qczc8ci0ylE?t=409

AlexMAK1 replied:
Ок, посмотрю видео и ознакомлюсь на будущее. Спасибо.

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/categories/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+
+@Slf4j
+@Service
+public class CategoryServiceImpl implements CategoryService {
```
Тоже тут можно аннотацию чтобы от конструктора избавиться

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/categories/CategoryService.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.List;
+
+public interface CategoryService {
```
Сделано хорошо, но вот лично моя позиция что сервисы если требуется принимают реквест dto а отдают ентити, и если надо превращаем в другие дто в контроллере, но кто то любит отдавать из сервисов сразу dto тут кому как привычнее. Поэтому тут не критика, а скорее мнение =)

AlexMAK1 replied:
ок, ронял. приму во внимание. спасибо.

AlexMAK1 replied:
понял)))

AlexMAK1 replied:
ок. понял, приму во внимание. спасибо.

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/compilations/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+
+@Slf4j
+@Service
+public class CompilationServiceImpl implements CompilationService {
```
Аннотацию чтобы без конструктора

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/compilations/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+    public CompilationDto create(NewCompilationDto newCompilationDto) {
+        Set<Event> events = new HashSet<>();
+        for (long id : newCompilationDto.getEvents()) {
+            Event event = eventRepository.getReferenceById(id);
```
Вот тут не будет ли слишком много запросов в БД? Нельзя ли сразу все достать, на подумать

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/compilations/model/Compilation.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
У нас есть требование от создателей задачи, что бы в модели не было методов equals и hashCode
Причины:
1. Студенты случай с хибернейтом не проходили
2. Нет причин его использовать,
3. Это может привести к ошибкам

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/events/EventController.java:
```diff
@@ -0,0 +1,115 @@
+
+@Slf4j
+@RestController
+public class EventController {
```
Тоже про конструктор

AlexMAK1 replied:
done

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/events/EventController.java:
```diff
@@ -0,0 +1,115 @@
+                                        @RequestParam(name = "size", defaultValue = "10") Integer size,
+                                        @RequestParam(name = "sort", required = false) String sort,
+                                        HttpServletRequest request) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
```
Такой паттерн используется несколько раз, не лучше ли объект один раз создать?

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on stat_service/src/main/java/ru/practicum/stat_service/stats/StatsRepository.java:
```diff
@@ -0,0 +1,16 @@
+import java.time.LocalDateTime;
+
+public interface StatsRepository extends JpaRepository<Hit, Long> {
+    @Query("select COUNT(id) from Hit where timestamp>=?1 AND timestamp<=?2 AND uri LIKE ?3")
```
Почему тут используется ?1 вместо имен переменных?

AlexMAK1 replied:
Хороший вопрос. Сравнивал с UserRepository там используется переменная. Здесь не получается реализация в другом виде.

krylov-sergey-sergeevich commented on stat_service/src/main/java/ru/practicum/stat_service/stats/StatsServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+        LocalDateTime startTime;
+        LocalDateTime endTime;
+        String app = "main_service";
```
лучше такие переменные выносить в статику и какой то утилитный класс который их все хранит или переменные приложения

AlexMAK1 replied:
done

krylov-sergey-sergeevich commented on stat_service/src/main/java/ru/practicum/stat_service/stats/StatsServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public List<ViewStats> getStatistics(String start, String end, List<String> uris, boolean unique)
+            throws UnsupportedEncodingException {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
```
Опять создаем форматер хотя он уже в этом приложение где-то подобный есть вроде как =)

AlexMAK1 replied:
done

AlexMAK1 committed:
fixed bugs after first review

AlexMAK1 committed:
Merge remote-tracking branch 'origin/develop' into develop

krylov-sergey-sergeevich commented on main_service/src/main/java/ru/practicum/main_service/events/EventServiceImpl.java:
```diff
@@ -33,26 +33,18 @@
     private final UserRepository userRepository;
-
-    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd' 'HH:mm:ss");
+    public static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
```
Хорошо что теперь везде эту переменную используете, но лучше такие в утилитные классы выносить, на будущее

krylov-sergey-sergeevich reviewed (approved):
Привет, вижу правки - зачет!

krylov-sergey-sergeevich reviewed (changes requested):
Привет, проверил код в git все ок и потом понял что забыл о моменте с 

"Меня смутило отсутствие схемы бд в файлах схемы, так и запланировано было?" - поэтому вопрос где схема?)

AlexMAK1 committed:
add schemes after second review

krylov-sergey-sergeevich reviewed (approved):
Вот теперь хорошо

AlexMAK1 committed:
add feature comments

AlexMAK1 committed:
feature comments is done

AlexMAK1 committed:
added validation in CommentServiceImpl,
changed path for endpoints in CommentController,
added table comments in README.md

AlexMAK1 committed:
deleted code for creation table comments for avoiding conflicts

AlexMAK1 committed:
the last changes

AlexMAK1 committed:
Merge pull request #2 from AlexMAK1/feature

add feature comments
